### PR TITLE
Flatten rev

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

This PR flattens rev/fun and finishes flattening the rev folder by creating the stan/math/rev.hpp header. This again touches a lot of files, but the changes are very simple. Only a few files required manual merging, the rest are renaming changes.

## Tests

/

## Side Effects

/

## Checklist

- [x] Math issue #937 

- [x] Copyright holder: Rok Češnovar

### CODE REVIEW

File: test/unit/math/rev/functor/integrate_ode_rk45_grad_test.cpp [hunk 1]
```diff
@@ -1,7 +1,4 @@
-#include <stan/math/rev/arr.hpp>
-// mat is needed to get operands_and_partials working with
-// std::vector<var> edges needed in coupled_ode_observer
-#include <stan/math/rev/mat.hpp>
```
Comment (mcol) [inline test/unit/math/rev/functor/integrate_ode_rk45_grad_test.cpp]: Thanks!

File: test/unit/math/rev/prob/expect_eq_diffs.hpp [hunk 2]
```diff
@@ -1,7 +1,7 @@
 #ifndef TEST_UNIT_DISTRIBUTION_EXPECT_EQ_DIFFS_HPP
 #define TEST_UNIT_DISTRIBUTION_EXPECT_EQ_DIFFS_HPP
```
Comment (mcol) [inline test/unit/math/rev/prob/expect_eq_diffs.hpp]: Pre-existing, but this header guard doesn't match the path to the file, and it's the only one with "DISTRIBUTION" in the name.

File: test/unit/math/rev/prob/inv_wishart2_test.cpp [hunk 3]
```diff
@@ -1,7 +1,7 @@
-#include <stan/math/rev/mat.hpp>
+#include <stan/math/rev.hpp>
 #include <gtest/gtest.h>
 #include <test/unit/math/rev/prob/expect_eq_diffs.hpp>
-#include <test/unit/math/rev/mat/util.hpp>
+#include <test/unit/math/rev/util.hpp>
```
Comment (mcol) [inline test/unit/math/rev/prob/inv_wishart2_test.cpp]: Put all rev headers together.

File: test/unit/math/rev/prob/multi_gp2_test.cpp [hunk 4]
```diff
@@ -1,9 +1,9 @@
-#include <stan/math/rev/mat.hpp>
+#include <stan/math/rev.hpp>
 #include <gtest/gtest.h>
 #include <test/unit/math/rev/prob/expect_eq_diffs.hpp>
 #include <test/unit/math/rev/prob/test_gradients.hpp>
 #include <test/unit/math/prim/prob/agrad_distributions_multi_gp.hpp>
-#include <test/unit/math/rev/mat/util.hpp>
+#include <test/unit/math/rev/util.hpp>
```
Comment (mcol) [inline test/unit/math/rev/prob/multi_gp2_test.cpp]: Put all rev headers together.

File: test/unit/math/rev/prob/multi_gp_cholesky2_test.cpp [hunk 5]
```diff
@@ -1,9 +1,9 @@
-#include <stan/math/rev/mat.hpp>
+#include <stan/math/rev.hpp>
 #include <gtest/gtest.h>
 #include <test/unit/math/rev/prob/expect_eq_diffs.hpp>
 #include <test/unit/math/rev/prob/test_gradients.hpp>
 #include <test/unit/math/prim/prob/agrad_distributions_multi_gp_cholesky.hpp>
-#include <test/unit/math/rev/mat/util.hpp>
+#include <test/unit/math/rev/util.hpp>
```
Comment (mcol) [inline test/unit/math/rev/prob/multi_gp_cholesky2_test.cpp]: Put all rev headers together.

File: test/unit/math/rev/prob/multi_student_t2_test.cpp [hunk 6]
```diff
@@ -1,10 +1,10 @@
-#include <stan/math/rev/mat.hpp>
+#include <stan/math/rev.hpp>
 #include <gtest/gtest.h>
 #include <test/unit/math/rev/prob/expect_eq_diffs.hpp>
 #include <test/unit/math/rev/prob/test_gradients.hpp>
 #include <test/unit/math/rev/prob/test_gradients_multi_student_t.hpp>
 #include <test/unit/math/prim/prob/agrad_distributions_multi_student_t.hpp>
-#include <test/unit/math/rev/mat/util.hpp>
+#include <test/unit/math/rev/util.hpp>
```
Comment (mcol) [inline test/unit/math/rev/prob/multi_student_t2_test.cpp]: All rev headers together.

File: test/unit/math/rev/prob/multinomial_test.cpp [hunk 7]
```diff
@@ -1,7 +1,7 @@
-#include <stan/math/rev/mat.hpp>
+#include <stan/math/rev.hpp>
 #include <gtest/gtest.h>
 #include <test/unit/math/rev/prob/expect_eq_diffs.hpp>
-#include <test/unit/math/rev/mat/util.hpp>
+#include <test/unit/math/rev/util.hpp>
```
Comment (mcol) [inline test/unit/math/rev/prob/multinomial_test.cpp]: All rev headers together.

File: test/unit/math/rev/prob/wishart_test.cpp [hunk 8]
```diff
@@ -1,7 +1,7 @@
-#include <stan/math/rev/mat.hpp>
+#include <stan/math/rev.hpp>
 #include <gtest/gtest.h>
 #include <test/unit/math/rev/prob/expect_eq_diffs.hpp>
-#include <test/unit/math/rev/mat/util.hpp>
+#include <test/unit/math/rev/util.hpp>
```
Comment (mcol) [inline test/unit/math/rev/prob/wishart_test.cpp]: All rev headers together.

**Comment by rok-cesnovar** [comment]
Thanks @mcol for the thorough reviews and quick turnaround. It really made this flattening much easier. It was never going to be painless but at least we get through it fast :)

**Comment by rok-cesnovar** [comment]
Also, I addressed all the comments.

**Comment by rok-cesnovar** [comment]
@mcol In order to fix the issue of the associated Stan PR (https://github.com/stan-dev/stan/pull/2875) I had to reverse the order of includes of prim/arr and prim/math. 

If we include arr first, the compiler is unable to compile fill with `vector<vector<eigen matrix>>`:
`no matching function for call to ‘fill(std::vector<std::vector<Eigen::Matrix<double, -1, -1> > >&, local_scalar_t__&)’`. 

If we include mat first, it compiles fine. 

The actual proper fix for this would be that https://github.com/stan-dev/math/blob/develop/stan/math/prim/arr/fun/fill.hpp would include mat/fun/fill.hpp. But that goes against the rules set for the arr/mat structure (make test-math-dependencies would fail). Before the flatten it worked because Stan included rev/mat.hpp and that file then includes arr.hpp. 

After prim gets flatten all this weirdness will go away.

**Pull request merged**